lightning/src/ln/features.rs

```diff
@@ -395,6 +399,68 @@ impl InvoiceFeatures {
 	}
 }
 
+impl ToBase32 for InvoiceFeatures {
+	fn write_base32<W: WriteBase32>(&self, writer: &mut W) -> Result<(), <W as WriteBase32>::Err> {
+		// Explanation for the "4": the normal way to round up when dividing is to add the divisor
+		// minus one before dividing
+		let total_u5s = (self.flags.len() * 8 + 4) / 5 as usize;
+		let mut be_u5s: Vec<u5> = vec![u5::try_from_u8(0).unwrap(); total_u5s];
+		for (u8_idx, byte) in self.flags.iter().enumerate() {
+			for bit_pos in 0..8 {
+				if 1 << bit_pos & byte != 0 {
+					let bit_pos_from_left_0_indexed = u8_idx * 8 + (bit_pos);
+					let new_u5_idx = total_u5s - (bit_pos_from_left_0_indexed / 5) as usize - 1;
+					let new_bit_pos = bit_pos_from_left_0_indexed % 5;
+					be_u5s[new_u5_idx] = u5::try_from_u8(be_u5s[new_u5_idx].to_u8() | 1 << new_bit_pos).unwrap();
+				}
+			}
+		}
+
+		// Trim the starting 0s (they're are the highest feature bits, so if they're unset they should
+		// be dropped)
+		while be_u5s.len() > 0 && be_u5s[0].to_u8() == 0 {
+			be_u5s.remove(0);
+		}
+		writer.write(&be_u5s)
+	}
+}
+
+impl Base32Len for InvoiceFeatures {
+	fn base32_len(&self) -> usize {
+		self.to_base32().len()
+	}
+}
+
+impl FromBase32 for InvoiceFeatures {
+	type Err = bech32::Error;
+
+	fn from_base32(field_data: &[u5]) -> Result<InvoiceFeatures, bech32::Error> {
+		// Explanation for the "7": the normal way to round up when dividing is to add the divisor
+		// minus one before dividing
+		let total_bytes = (field_data.len() * 5 + 7) / 8 as usize;
+		let mut le_bytes: Vec<u8> = vec![0; total_bytes];
+		for (u5_idx, chunk) in field_data.iter().enumerate() {
+			let chunk_as_u8 = chunk.to_u8();
+			for bit_pos in 0..5 {
```

Comment:
Same here - I don't think its acceptable to branch on each bit in the feature while reading.